crates/nargo/src/workspace.rs

```diff
@@ -2,3 +2,79 @@
 // Then we use workspace to allow more than one. In the future, do not allow there to be
 // both a binary and a library.
 // - library will be default
+
+use std::{
+    iter::{once, Once},
+    path::PathBuf,
+    slice,
+};
+
+use noirc_frontend::graph::CrateName;
+
+use crate::{
+    constants::{CONTRACT_DIR, PROOFS_DIR, TARGET_DIR},
+    package::Package,
+};
+
+#[derive(Clone)]
+pub struct Workspace {
+    pub root_dir: PathBuf,
+    pub members: Vec<Package>,
+    // Selected is --package or default
+    pub selected_package: Option<CrateName>,
+}
+
+impl Workspace {
+    pub fn package_build_path(&self, package: &Package) -> Option<PathBuf> {
+        package
+            .name
+            .as_ref()
+            .map(|crate_name| self.target_directory_path().join(crate_name.as_string()))
+    }
+
+    pub fn contracts_directory_path(&self) -> PathBuf {
+        self.root_dir.join(CONTRACT_DIR)
+    }
+
+    pub fn proofs_directory_path(&self) -> PathBuf {
+        self.root_dir.join(PROOFS_DIR)
+    }
+
+    pub fn target_directory_path(&self) -> PathBuf {
+        self.root_dir.join(TARGET_DIR)
+    }
+}
+
+pub enum IntoIter<'a, T> {
+    Only(Once<&'a T>),
+    All(slice::Iter<'a, T>),
+}
+
+impl<'a> IntoIterator for &'a Workspace {
+    type Item = &'a Package;
+    type IntoIter = IntoIter<'a, Package>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        if let Some(selected_name) = &self.selected_package {
+            // Precondition: The selected_package needed to be verified when creating the workspace
```

Comment:
TODO: Need to make `Workspace::new()` that can return a result if the preconditions aren't fulfilled

Reply:
Doing this inside the manifest resolution function itself